src/System.IO.Pipelines.Networking.Tls/ApplicationProtocols.cs

```diff
@@ -0,0 +1,157 @@
+        {
+            None = 0,
+            Success,
+            SelectedClientOnly
+        }
+
+        internal static unsafe ProtocolIds FindNegotiatedProtocol(SSPIHandle context)
+        {
+            ContextApplicationProtocol protoInfo;
+            Interop.QueryContextAttributesW(ref context, ContextAttribute.ApplicationProtocol, out protoInfo);
+
+            if (protoInfo.ProtoNegoStatus != ApplicationProtocolNegotiationStatus.Success)
+            {
+                throw new InvalidOperationException("Could not negotiate a mutal application protocol");
+            }
+            return GetNegotiatedProtocol(protoInfo.ProtocolId, protoInfo.ProtocolIdSize);
+        }
+
+        public static byte[] GetBufferForProtocolId(ProtocolIds supportedProtocols, bool withAlpnHeader)
+        {
+            short listLength = 0;
+            for (int i = 0; i < _numberOfProtocols.Length; i++)
+            {
+                if (((int) _numberOfProtocols.GetValue(i) & (int) supportedProtocols) > 0)
+                {
+                    listLength += (short) (_allProtocols[i].Length + 1);
+                }
+            }
+            byte[] returnValue;
+            if (withAlpnHeader)
+            {
+                //Now we know the total length of the list lets create our array of protocols and header
+                returnValue = new byte[listLength + _HeaderLength];
+                var spa = new Span<byte>(returnValue);
+                int length = _HeaderLength + listLength - _ContentOffset;
+                spa.Write(length);
+                spa = spa.Slice(4);
+                spa.Write(ApplicaitonProtocolNegotiationExtension.Alpn);
+                spa = spa.Slice(4);
+                spa.Write(listLength);
+                spa = spa.Slice(2);
+                ProtocolList(supportedProtocols, spa);
+            }
+            else
+            {
+                //No header just make an array of the correct size
+                returnValue = new byte[listLength];
+                var spa = new Span<byte>(returnValue);
+                ProtocolList(supportedProtocols, spa);
+            }
+            return returnValue;
+        }
+
+        private static Span<byte> ProtocolList(ProtocolIds supportedProtocols, Span<byte> spa)
+        {
+            for (int i = 0; i < _numberOfProtocols.Length; i++)
+            {
+                if (((int) _numberOfProtocols.GetValue(i) & (int) supportedProtocols) > 0)
+                {
+                    var value = _allProtocols[i];
+                    spa.Write((byte) value.Length);
+                    spa = spa.Slice(1);
+                    var valueSpan = new Span<byte>(value);
+                    valueSpan.CopyTo(spa);
+                    spa = spa.Slice(value.Length);
+                }
+            }
+
+            return spa;
+        }
+
+        internal static unsafe ProtocolIds GetNegotiatedProtocol(byte* protocolId, byte protocolIdSize)
```

Comment:
This should just take the span, to keep the pointer and size as close together as possible.  (Especially since all it does with the pointer is Span it)